src/AgentVisibilityContext.tsx

```diff
@@ -1,90 +1,71 @@
+import { VisibilitySelectionMap } from './constants';
+import { getNewSelectionMap, mapUIDisplayDataToSelectionMap } from './utils';
+import { SelectionType } from './types';
 
 interface VisibilityContextType {
   uiDisplayData: UIDisplayData;
   hiddenAgents: VisibilitySelectionMap;
-  allAgentsHidden: VisibilitySelectionMap;
-  noAgentsHidden: VisibilitySelectionMap;
+  highlightedAgents: VisibilitySelectionMap;
   receiveUIDisplayData: (data: UIDisplayData) => void;
-  handleAllAgentsCheckboxChange: (hiddenState: CheckboxState) => void;
-  handleAgentCheckboxChange: (agentName: string) => void;
+  setHiddenAgents: React.Dispatch<React.SetStateAction<VisibilitySelectionMap>>;
+  toggleAgentVisibility: (
+    agentName: string,
+    selectionType: SelectionType
+  ) => void;
 }
 
 export const VisibilityContext = createContext<VisibilityContextType>({
   uiDisplayData: [],
   hiddenAgents: {},
-  allAgentsHidden: {},
-  noAgentsHidden: {},
+  highlightedAgents: {},
   receiveUIDisplayData: () => {},
-  handleAllAgentsCheckboxChange: () => {},
-  handleAgentCheckboxChange: () => {},
+  setHiddenAgents: () => {},
+  toggleAgentVisibility: () => {},
 });
 
-export const VisibilityProvider = ({
-  children,
-}: {
-  children: React.ReactNode;
-}) => {
-  const [uiDisplayData, setUiDisplayData] = React.useState<UIDisplayData>([]);
-  const [hiddenAgents, setHiddenAgents] =
-    React.useState<VisibilitySelectionMap>({});
-
-  /**
-   * noAgentsHidden maps each agent name to a value of [agent]
-   * allAgentsHidden maps each agent name to an empty array
-   * They serve as values for the handler of the hide all agents checkbox,
-   * and for checks to determine state of that checkbox.
-   * They are memoized to prevent unnecessary re-renders, and
-   * set when the uiDisplayData arrives.
-   */
-  const noAgentsHidden: VisibilitySelectionMap = useMemo(() => {
-    return uiDisplayData.reduce<VisibilitySelectionMap>((acc, item) => {
-      acc[item.name] = [item.name];
-      return acc;
-    }, {});
-  }, [uiDisplayData]);
-
-  const allAgentsHidden: VisibilitySelectionMap = useMemo(() => {
-    return uiDisplayData.reduce<VisibilitySelectionMap>((acc, item) => {
-      acc[item.name] = [];
-      return acc;
-    }, {});
-  }, [uiDisplayData]);
-
-  useEffect(() => {
-    setHiddenAgents(noAgentsHidden);
-  }, [noAgentsHidden]);
+export const VisibilityProvider = ({ children }: { children: ReactNode }) => {
+  const [uiDisplayData, setUiDisplayData] = useState<UIDisplayData>([]);
+  const [hiddenAgents, setHiddenAgents] = useState<VisibilitySelectionMap>({});
+  const [highlightedAgents, setHighlightedAgents] =
+    useState<VisibilitySelectionMap>({});
 
   const receiveUIDisplayData = (data: UIDisplayData) => {
     setUiDisplayData(data);
+    const noAgentsSelectedMap = mapUIDisplayDataToSelectionMap(data);
+    setHighlightedAgents(noAgentsSelectedMap);
+    setHiddenAgents(noAgentsSelectedMap);
```

Comment:
@meganrm I know you wanted this call of `setHiddenAgents` to come from the checkbox but it just seems more readable and less confusing to have these two calls in the same place, and not have a useEffect in the checkbox component

Reply:
that's fine. one suggestion you could take or leave: make a function called "setInitialSelections" or "initializeSelectionUI" that just wraps these three lines of code. then it would be clear the two things you do when you get new ui data is 1.save it, 2. initialize the ui.